File: labs/M19-tracing-logging/starter/instrumenter.js
/**
 * M19 Lab — Agent Instrumenter (Starter)
 * =======================================
 * Wrap a UCC research agent with automatic tracing and structured
 * logging. Every API call and tool execution gets its own span.
 *
 * KEY CONCEPT: Instrumentation means adding measurement code around
 * existing operations WITHOUT changing the operations themselves.
 *
 * Usage:
 *     node instrumenter.js
 */

import { Trace, Span, withSpan } from "./trace_model.js";
import { StructuredLogger, scrubPii } from "./structured_logger.js";

// =============================================================================
// MOCK AGENT
// =============================================================================

// WHAT: A fake agent that simulates a real Claude-powered UCC research agent.
// WHY:  Realistic agent to instrument without needing an API key.

class MockUCCAgent {
  constructor() {
    this.model = "claude-sonnet-4-20250514";
    this.callLog = [];
  }

  async run(query) {
    const operations = [];

    // Step 1: First LLM call
    await new Promise((r) => setTimeout(r, 50));
    operations.push({
      type: "llm_call",
      model: this.model,
      inputTokens: 350,
      outputTokens: 85,
      durationMs: 50.0,
      response: {
        stopReason: "tool_use",
        toolName: "search_filings",
        toolInput: { debtor_name: "Greenfield Logistics", state: "New York" },
      },
    });

    // Step 2: Tool execution
    await new Promise((r) => setTimeout(r, 20));
    operations.push({
      type: "tool_call",
      toolName: "search_filings",
      toolInput: { debtor_name: "Greenfield Logistics", state: "New York" },
      toolOutput: {
        results: [
          {
            filing_number: "NY-2024-001234",
            debtor: "Greenfield Logistics LLC",
            secured_party: "First National Bank",
            collateral: "All inventory and equipment",
            filing_date: "2024-03-15",
            status: "active",
          },
          {
            filing_number: "NY-2024-005678",
            debtor: "Greenfield Logistics LLC",
            secured_party: "Atlas Capital Partners",
            collateral: "Accounts receivable",
            filing_date: "2024-07-22",
            status: "active",
          },
        ],
      },
      durationMs: 20.0,
    });

    // Step 3: Second LLM call
    await new Promise((r) => setTimeout(r, 30));
    const answerText =
      "I found 2 active UCC filings for Greenfield Logistics LLC in New York:\n\n" +
      "1. NY-2024-001234 \u2014 Filed 2024-03-15 by First National Bank against all inventory and equipment.\n" +
      "2. NY-2024-005678 \u2014 Filed 2024-07-22 by Atlas Capital Partners against accounts receivable.\n\n" +
      "Both filings are currently active.";

    operations.push({
      type: "llm_call",
      model: this.model,
      inputTokens: 820,
      outputTokens: 210,
      durationMs: 30.0,
      response: { stopReason: "end_turn", text: answerText },
    });

    this.callLog = operations;
    return { answer: answerText, operations };
  }
}

// =============================================================================
// INSTRUMENTED AGENT
// =============================================================================

class InstrumentedAgent {
  /**
   * @param {object} agent - Any object with a .run(query) method
   * @param {StructuredLogger} logger
   */
  constructor(agent, logger) {
    // TODO: Store agent and logger as instance properties
  }

  /**
   * Execute the agent with full tracing.
   * @returns {Promise<[object, Trace]>} [result, trace]
   */
  async run(query) {
    // TODO: Implement instrumented agent execution:
    //
    // 1. Create a new Trace with name "agent_request"
    //    and metadata { query }
    //
    // 2. Use withSpan to create the root span "agent_request":
    //    await withSpan(trace, "agent_request", null, async (root) => { ... })
    //
    // 3. Inside the callback:
    //    a. Set attribute "query" on root
    //    b. Log "Agent request started"
    //    c. const result = await this.agent.run(query)
    //    d. For each operation in result.operations:
    //       - "llm_call" -> await this._instrumentLlmCall(trace, root, op)
    //       - "tool_call" -> await this._instrumentToolCall(trace, root, op)
    //    e. Set attribute "answer_length" on root
    //    f. Log "Agent request completed"
    //    g. Return result
    //
    // 4. Return [result, trace]
  }

  async _instrumentLlmCall(trace, parent, operation) {
    // TODO:
    // 1. Use withSpan to create "llm_call" child span
    // 2. Set attributes: model, inputTokens, outputTokens, totalTokens, stopReason
    // 3. Log with this.logger.logLlmCall()
    // 4. Sleep for operation.durationMs
  }

  async _instrumentToolCall(trace, parent, operation) {
    // TODO:
    // 1. Use withSpan to create "tool_execution" child span
    // 2. Set attributes: toolName, toolInput, resultCount
    // 3. Log with this.logger.logToolCall()
    // 4. Sleep for operation.durationMs
  }
}

// =============================================================================
// SELF-TEST
// =============================================================================

async function selfTest() {
  console.log("=".repeat(60));
  console.log("M19 Agent Instrumenter \u2014 Self-Test");
  console.log("=".repeat(60));
  console.log();

  const logger = new StructuredLogger("ucc_agent");
  const agent = new MockUCCAgent();
  const instrumented = new InstrumentedAgent(agent, logger);

  const [result, trace] = await instrumented.run(
    "Find all UCC filings for Greenfield Logistics in New York"
  );

  console.log(`\n${"=".repeat(60)}`);
  console.log("TRACE SUMMARY");
  console.log("=".repeat(60));
  console.log(`  Trace ID:    ${trace.traceId}`);
  console.log(`  Total Spans: ${trace.spans.length}`);
  console.log(`  Duration:    ${trace.getDurationMs().toFixed(1)} ms`);
  console.log();

  for (const span of trace.spans) {
    const indent = span.parentSpanId ? "    " : "  ";
    const icon = span.status === "ok" ? "OK" : "ERR";
    console.log(`${indent}[${icon}] ${span.name} \u2014 ${span.durationMs.toFixed(1)} ms`);
    if (Object.keys(span.attributes).length > 0) {
      for (const [k, v] of Object.entries(span.attributes)) {
        console.log(`${indent}     ${k}: ${JSON.stringify(v)}`);
      }
    }
    console.log();
  }

  console.log("=".repeat(60));
  console.log("AGENT ANSWER");
  console.log("=".repeat(60));
  console.log(result.answer);
  console.log();

  console.assert(trace.spans.length === 4, `Expected 4 spans, got ${trace.spans.length}`);
  const spanNames = trace.spans.map((s) => s.name);
  console.assert(spanNames.includes("agent_request"), "Missing root span");
  console.assert(spanNames.filter((n) => n === "llm_call").length === 2, "Expected 2 llm_call spans");
  console.assert(spanNames.includes("tool_execution"), "Missing tool_execution span");
  console.log("All assertions passed!");
}

export { MockUCCAgent, InstrumentedAgent };

const isMain = process.argv[1] && process.argv[1].endsWith("instrumenter.js");
if (isMain) {
  selfTest().catch(console.error);
}

File: labs/M19-tracing-logging/starter/instrumenter.py
"""
M19 Lab — Agent Instrumenter (Starter)
=======================================
Wrap a UCC research agent with automatic tracing and structured
logging. Every API call and tool execution gets its own span.

KEY CONCEPT: Instrumentation means adding measurement code around
existing operations WITHOUT changing the operations themselves.
Think of it like wrapping a gift — the gift (your agent logic)
stays the same, but now it's wrapped in timing and logging code
that makes it observable.

Usage:
    python instrumenter.py
"""

import time
import json
import sys
import os

# Import our trace model and logger from the same directory
sys.path.insert(0, os.path.dirname(__file__))
from trace_model import Trace, Span, SpanContext
from structured_logger import StructuredLogger, scrub_pii


# =============================================================================
# MOCK AGENT — simulates a real agent without needing an API key
# =============================================================================

# WHAT: A fake agent that simulates the exact sequence of operations
#   a real Claude-powered UCC research agent would perform:
#   1. Receive a query
#   2. Call Claude (LLM call #1) to decide which tool to use
#   3. Execute the tool (search_filings)
#   4. Call Claude (LLM call #2) to synthesize the answer
# WHY:  You need a realistic agent to instrument, but you don't want
#   to burn API credits or require an API key for a tracing lab.
# GOTCHA: The mock uses time.sleep() to simulate realistic latencies.
#   In production, these would be real network calls.

class MockUCCAgent:
    """Simulates a UCC research agent: 2 LLM calls + 1 tool call."""

    def __init__(self):
        self.model = "claude-sonnet-4-20250514"
        self.call_log = []  # Records every operation for verification

    def run(self, query: str) -> dict:
        """
        Execute a mock agent run. Returns a dict with:
        - "answer": the final text response
        - "operations": list of dicts describing each operation
        """
        operations = []

        # --- Step 1: First LLM call (decide which tool to use) ---
        time.sleep(0.05)  # simulate API latency
        llm_call_1 = {
            "type": "llm_call",
            "model": self.model,
            "input_tokens": 350,
            "output_tokens": 85,
            "duration_ms": 50.0,
            "response": {
                "stop_reason": "tool_use",
                "tool_name": "search_filings",
                "tool_input": {"debtor_name": "Greenfield Logistics", "state": "New York"}
            }
        }
        operations.append(llm_call_1)

        # --- Step 2: Tool execution ---
        time.sleep(0.02)  # simulate tool latency
        tool_call = {
            "type": "tool_call",
            "tool_name": "search_filings",
            "tool_input": {"debtor_name": "Greenfield Logistics", "state": "New York"},
            "tool_output": {
                "results": [
                    {
                        "filing_number": "NY-2024-001234",
                        "debtor": "Greenfield Logistics LLC",
                        "secured_party": "First National Bank",
                        "collateral": "All inventory and equipment",
                        "filing_date": "2024-03-15",
                        "status": "active"
                    },
                    {
                        "filing_number": "NY-2024-005678",
                        "debtor": "Greenfield Logistics LLC",
                        "secured_party": "Atlas Capital Partners",
                        "collateral": "Accounts receivable",
                        "filing_date": "2024-07-22",
                        "status": "active"
                    }
                ]
            },
            "duration_ms": 20.0
        }
        operations.append(tool_call)

        # --- Step 3: Second LLM call (synthesize the answer) ---
        time.sleep(0.03)  # simulate API latency
        llm_call_2 = {
            "type": "llm_call",
            "model": self.model,
            "input_tokens": 820,
            "output_tokens": 210,
            "duration_ms": 30.0,
            "response": {
                "stop_reason": "end_turn",
                "text": (
                    "I found 2 active UCC filings for Greenfield Logistics LLC "
                    "in New York:\n\n"
                    "1. NY-2024-001234 — Filed 2024-03-15 by First National Bank "
                    "against all inventory and equipment.\n"
                    "2. NY-2024-005678 — Filed 2024-07-22 by Atlas Capital Partners "
                    "against accounts receivable.\n\n"
                    "Both filings are currently active."
                )
            }
        }
        operations.append(llm_call_2)

        self.call_log = operations
        return {
            "answer": llm_call_2["response"]["text"],
            "operations": operations
        }


# =============================================================================
# INSTRUMENTED AGENT
# =============================================================================

# WHAT: A wrapper that takes any agent function, creates a trace for
#   each run, and wraps every operation in a measured span.
# WHY:  Without instrumentation, you have no idea how long each step
#   takes, which tokens were used, or where errors occurred. The
#   instrumenter adds this visibility automatically.
# GOTCHA: The instrumenter should NOT change the agent's behavior —
#   it only adds measurement. If the agent would return X without
#   instrumentation, it must still return X with it.

class InstrumentedAgent:
    """Wraps an agent with automatic tracing and structured logging."""

    def __init__(self, agent, logger: StructuredLogger):
        """
        Args:
            agent: Any object with a .run(query) method
            logger: StructuredLogger instance for JSON log output
        """
        # TODO: Store agent and logger as instance attributes
        pass

    def run(self, query: str) -> tuple:
        """
        Execute the agent with full tracing.

        Returns:
            (result, trace) — the agent's result and the complete trace
        """
        # TODO: Implement instrumented agent execution:
        #
        # 1. Create a new Trace with name="agent_request"
        #    and metadata={"query": query}
        #
        # 2. Use SpanContext to create the root span "agent_request"
        #    Set attribute "query" on the root span
        #    Log an INFO message "Agent request started"
        #
        # 3. Inside the root span context:
        #    a. Call self.agent.run(query) to get the result
        #    b. Iterate over result["operations"]
        #    c. For each operation, call the appropriate _instrument_* method
        #       based on operation["type"]:
        #       - "llm_call" -> self._instrument_llm_call(trace, root, operation)
        #       - "tool_call" -> self._instrument_tool_call(trace, root, operation)
        #    d. Set attribute "answer_length" on root span = len(result["answer"])
        #    e. Log an INFO message "Agent request completed"
        #
        # 4. If any exception occurs:
        #    a. Log the error with self.logger.log_error()
        #    b. Re-raise the exception
        #
        # 5. Return (result, trace)
        pass

    def _instrument_llm_call(self, trace: Trace, parent: Span, operation: dict) -> None:
        """Create a span for an LLM API call."""
        # TODO:
        # 1. Use SpanContext to create a child span named "llm_call"
        #    under parent
        # 2. Set these attributes on the span:
        #    - "model": operation["model"]
        #    - "input_tokens": operation["input_tokens"]
        #    - "output_tokens": operation["output_tokens"]
        #    - "total_tokens": input + output
        #    - "stop_reason": operation["response"]["stop_reason"]
        # 3. Log the LLM call using self.logger.log_llm_call()
        #    with trace_id and span_id
        # 4. Sleep for operation["duration_ms"] / 1000 to simulate timing
        pass

    def _instrument_tool_call(self, trace: Trace, parent: Span, operation: dict) -> None:
        """Create a span for a tool execution."""
        # TODO:
        # 1. Use SpanContext to create a child span named "tool_execution"
        #    under parent
        # 2. Set these attributes on the span:
        #    - "tool_name": operation["tool_name"]
        #    - "tool_input": operation["tool_input"]
        #    - "result_count": number of results in tool_output["results"]
        # 3. Log the tool call using self.logger.log_tool_call()
        #    with trace_id and span_id
        # 4. Sleep for operation["duration_ms"] / 1000 to simulate timing
        pass


# =============================================================================
# SELF-TEST
# =============================================================================

def self_test():
    """Run the mock agent with full instrumentation."""
    print("=" * 60)
    print("M19 Agent Instrumenter — Self-Test")
    print("=" * 60)
    print()

    # Create logger and mock agent
    logger = StructuredLogger(service_name="ucc_agent")
    agent = MockUCCAgent()
    instrumented = InstrumentedAgent(agent, logger)

    # Run the instrumented agent
    result, trace = instrumented.run("Find all UCC filings for Greenfield Logistics in New York")

    # Print trace summary
    print(f"\n{'=' * 60}")
    print(f"TRACE SUMMARY")
    print(f"{'=' * 60}")
    print(f"  Trace ID:    {trace.trace_id}")
    print(f"  Total Spans: {len(trace.spans)}")
    print(f"  Duration:    {trace.get_duration_ms():.1f} ms")
    print()

    for span in trace.spans:
        indent = "    " if span.parent_span_id else "  "
        status_icon = "OK" if span.status == "ok" else "ERR"
        print(f"{indent}[{status_icon}] {span.name} — {span.duration_ms:.1f} ms")
        if span.attributes:
            for k, v in span.attributes.items():
                print(f"{indent}     {k}: {v}")
        print()

    # Print the answer
    print(f"{'=' * 60}")
    print(f"AGENT ANSWER")
    print(f"{'=' * 60}")
    print(result["answer"])
    print()

    # Verify
    assert len(trace.spans) == 4, f"Expected 4 spans (root + 2 LLM + 1 tool), got {len(trace.spans)}"
    span_names = [s.name for s in trace.spans]
    assert "agent_request" in span_names, "Missing root span"
    assert span_names.count("llm_call") == 2, "Expected 2 llm_call spans"
    assert "tool_execution" in span_names, "Missing tool_execution span"
    print("All assertions passed!")


if __name__ == "__main__":
    self_test()
